core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -241,6 +225,63 @@ impl ZkSyncStateKeeper {
         Err(Error::Canceled)
     }
 
+    /// This function is meant to be called only once during the state-keeper initialization.
+    /// It will check if we should load a protocol upgrade or a `setChainId` transaction,
+    /// perform some checks and return it.
+    pub(super) async fn load_protocol_upgrade_tx(
+        &mut self,
+        pending_miniblocks: &[MiniblockExecutionData],
+        protocol_version: ProtocolVersionId,
+        l1_batch_number: L1BatchNumber,
+    ) -> Result<Option<ProtocolUpgradeTx>, Error> {
+        // After the Shared Bridge is integrated,
+        // there has to be a setChainId upgrade transaction after the chain genesis.
+        // It has to be the first transaction of the first batch.
+        // The setChainId upgrade does not bump the protocol version, but attaches an upgrade
+        // transaction to the genesis protocol version version.
+        let first_batch_in_shared_bridge =
+            l1_batch_number == L1BatchNumber(1) && !protocol_version.is_pre_shared_bridge();
+        let previous_batch_protocol_version =
+            self.io.load_previous_batch_version_id().await.unwrap();
+
+        let version_changed = protocol_version != previous_batch_protocol_version;
+        let protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
+
+        let protocol_upgrade_tx = if pending_miniblocks.is_empty()
+            && (version_changed || first_batch_in_shared_bridge)
+        {
+            // We have a new upgrade transaction - either a regular protocol upgrade or a `setChainId` upgrade.
+            // If we are running an EN, `protocol_upgrade_tx` will be `None` here since it is fetched from the main node.
+            tracing::info!("There is an new upgrade tx to be executed in this batch");
```

Comment:
Nit: Here and in other log messages in this method, it could make sense to add context (e.g., "this batch" -> "batch #{l1_batch_number}"; info about `protocol_upgrade_tx`) to simplify debugging.